osquery/filesystem/filesystem.cpp

```diff
@@ -82,18 +82,28 @@ Status writeTextFile(const fs::path& path,
 
 struct OpenReadableFile : private boost::noncopyable {
  public:
-  explicit OpenReadableFile(const fs::path& path, bool blocking = false) {
+  explicit OpenReadableFile(const fs::path& path, bool blocking = false)
+      : blocking_io(blocking) {
     int mode = PF_OPEN_EXISTING | PF_READ;
     if (!blocking) {
       mode |= PF_NONBLOCK;
     }
 
     // Open the file descriptor and allow caller to perform error checking.
-    fd.reset(new PlatformFile(path, mode));
+    fd = std::make_unique<PlatformFile>(path, mode);
+
+    if (!blocking && fd->isSpecialFile()) {
+      // A special file cannot be read in non-blocking mode, reopen in blocking
+      // mode
+      mode &= ~PF_NONBLOCK;
+      blocking_io = true;
+      fd = std::make_unique<PlatformFile>(path, mode);
```

Comment:
Is there some reason we open the file twice, vs performing `fd = std::make_unique<PlatformFile>(path, mode);` once after this if-block?

Reply:
The problem here is that isSpecialFile, which is used in the if, is a function of PlatformFile, which needs an instance of it to work (especially on Windows where there's actually a handle to open), so we need to open it in a mode and then change it later.